src/Mod/Sketcher/Gui/SnapManager.cpp

```diff
@@ -0,0 +1,223 @@
+        snapToGrid(x, y);
+    }
+}
+
+void SnapManager::snapAtAngle(double& x, double& y)
+{
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+    double snapAngle = hGrp->GetFloat("SnapAngle", 5.) * M_PI / 180;
+
+    Base::Vector2d pointToOverride(x, y);
+    double length = (pointToOverride - angleSnapRef).Length();
+    double angle = (pointToOverride - angleSnapRef).Angle();
+    angle = round(angle / snapAngle) * snapAngle;
+    pointToOverride = angleSnapRef + length * Base::Vector2d(cos(angle), sin(angle));
+    x = pointToOverride.x;
+    y = pointToOverride.y;
+}
+
+bool SnapManager::snapToObject(double& x, double& y)
+{
+    Sketcher::SketchObject* Obj = viewProvider.getSketchObject();
+    int geoId = GeoEnum::GeoUndef;
+    Sketcher::PointPos posId = Sketcher::PointPos::none;
+
+    int VtId = ViewProviderSketchSnapAttorney::getPreselectPoint(viewProvider);
+    int CrsId = ViewProviderSketchSnapAttorney::getPreselectCross(viewProvider);
+    int CrvId = ViewProviderSketchSnapAttorney::getPreselectCurve(viewProvider);
+
+    if (CrsId == 0 || VtId >= 0) {
+        if (CrsId == 0) {
+            geoId = Sketcher::GeoEnum::RtPnt;
+            posId = Sketcher::PointPos::start;
+        }
+        else if (VtId >= 0) {
+            Obj->getGeoVertexIndex(VtId, geoId, posId);
+        }
+
+        x = Obj->getPoint(geoId, posId).x;
+        y = Obj->getPoint(geoId, posId).y;
+        return true;
+    }
+    else if (CrsId == 1) { //H_Axis
+        y = 0;
+        return true;
+    }
+    else if (CrsId == 2) { //V_Axis
+        x = 0;
+        return true;
+    }
+    else if (CrvId >= 0 || CrvId <= Sketcher::GeoEnum::RefExt) { //Curves
+
+        const Part::Geometry* geo = Obj->getGeometry(CrvId);
+
+        Base::Vector3d pointToOverride(x, y, 0.);
+
+        double pointParam;
+        auto curve = static_cast<const Part::GeomCurve*>(geo);
+        try {
+            curve->closestParameter(pointToOverride, pointParam);
+            pointToOverride = curve->pointAtParameter(pointParam);
+        }
+        catch (Base::CADKernelError& e) {
+            e.ReportException();
+            return false;
+        }
+
+        //If it is a line, then we check if we need to snap to the middle.
+        if (geo->getTypeId() == Part::GeomLineSegment::getClassTypeId()) {
+            snapToLineMiddle(pointToOverride, geo);
+        }
+
+        //If it is an arc, then we check if we need to snap to the middle (not the center).
+        if (geo->getTypeId() == Part::GeomArcOfCircle::getClassTypeId()) {
+            snapToArcMiddle(pointToOverride, geo);
+        }
+
+        x = pointToOverride.x;
+        y = pointToOverride.y;
+
+        return true;
+    }
+
+    return false;
+}
+
+void SnapManager::snapToGrid(double& x, double& y)
+{
+    // Snap Tolerance in pixels
+    const double snapTol = viewProvider.GridSize.getValue() / 5;

```

Comment:
Shouldn't this be a parameter? No need for a settings variable in my opinion, but this is a quite course snapping (20% of grid size), especially for large grid sizes..

Same with line-> middle and arc->middle, both of those could perhaps share a modifiable parameter?

Reply:
Thanks for the review ! 
For the grid snapping, I haven't changed this function, I just copy-pasted it. Personally I thought the 20% was behaving pretty well. But I could add a preference here. However I don't know if I need to change that in this PR though, as keeping things simple is better to merge efficiently.

For the line-middle and arc-middle, I can add them as parameters but I don't know if that is not bloating the preferences? I initially thought of using pick-radius so that the line-arc middles are acting as virtual points. I thought using a % was easier to snap to middle especially if you're zoomed in on the line. But maybe using pick radius would be better.

I will let Abdullah decide about what needs to be added now or later.

Reply:
We have three options here:
(a) do nothing
(b) create a const double local variable toleranceRatio = 0.2
(c) make it a parameter

For now I would go with (a) or (b). (b) is probably more elegant because it separates the parameter from the calculation. If this will be refactored in the future, not having to tweak calculations is always better. However, you will find many other places where (a) is done.

I do not think it is worth a preference ATM. There will always be a tradeoff between how much we put in preferences. I would like to see feedback from the user point of view first.